roqoqo/src/circuitdag.rs

```diff
@@ -0,0 +1,985 @@
+    /// The scope of the check is the whole graph.
+    ///
+    /// # Arguments:
+    ///
+    /// * `already_executed` - Slice of NodeIndices of Nodes that have already been executed in the Circuit.
+    /// * `to_be_executed` - NodeIndex of the Operation that should be executed next.
+    ///
+    /// # Returns:
+    ///
+    /// * `Vec<NodeIndex<usize>>` - Vector containing the sorted blocking elements.
+    pub fn execution_blocked(
+        &self,
+        already_executed: &[NodeIndex<usize>],
+        to_be_executed: &NodeIndex<usize>,
+    ) -> Vec<NodeIndex<usize>> {
+        let mut blocking_elements: Vec<NodeIndex<usize>> = vec![];
+        let mut rev_graph: Graph<Operation, (), Directed, usize> = self.graph.clone();
+        rev_graph.reverse();
+        let mut dfs = Dfs::new(&rev_graph, (*to_be_executed).into());
+        dfs.next(&rev_graph);
+
+        // Perform a DFS on the reversed graph starting from to_be_executed,
+        //  pushing on blocking_elements the nodes not contained in already_executed
+        while let Some(nxt) = dfs.next(&rev_graph) {
+            if !already_executed.contains(&nxt.index()) {
+                blocking_elements.push(nxt.index());
+            }
+        }
+
+        blocking_elements.sort_unstable();
+        blocking_elements
+    }
+
+    /// Checks which of the direct predecessors of an Operation in the CircuitDag blocks the execution.
+    ///
+    /// # Warning
+    ///
+    /// This method can only be used to determine if an operation can be executed when `already_executed` is consistent.
+    /// When the list  `already_executed` is inconsistent (a n operation is reported as executed that could not have been executed yet)
+    /// this method returning an empty vector does not imply that the `to_be_executed` operation can be executed.
+    ///
+    /// # Arguments:
+    ///
+    /// * `already_executed` - Slice of NodeIndices of Nodes that have already been executed in the Circuit.
+    /// * `to_be_executed` - NodeIndex of the Operation that should be executed next.
+    pub fn blocking_predecessors(
+        &self,
+        already_executed: &[NodeIndex<usize>],
+        to_be_executed: &NodeIndex<usize>,
+    ) -> Vec<NodeIndex<usize>> {
+        let mut blocking_elements: Vec<NodeIndex<usize>> = vec![];
+        let mut neighbor_iter = self
+            .graph
+            .neighbors_directed((*to_be_executed).into(), Incoming);
+
+        while let Some(nxt) = neighbor_iter.next() {
+            if !already_executed.contains(&nxt.index()) {
+                blocking_elements.push(nxt.index());
+            }
+        }
+
+        blocking_elements.sort_unstable();
+        blocking_elements
+    }
+
+    /// Returns a new front-layer after executing an operation from the current front layer.
+    ///
+    /// Returns an error if operation to be executed is not in the current front layer.
+    ///
+    /// # Arguments:
```

Comment:
We could describe in more detailed what a front layer is. We can say that a front-layer is a set of operations that are ready to be executed, because all required predecessors in the graph have already been executed (or also put in a link to an arxiv paper)